low-level/framework/utils/ipmi_client.py

```diff
@@ -161,39 +185,67 @@ def get_fru_list_by_type(self, fru_list, sensor_id_map):
-        result = process.communicate()
-        return result, process.returncode
+    def _run_ipmitool_subcommand(self, subcommand, grep_args=None):
+        """Executes ipmitool sub-commands, and optionally greps the output."""
+        self.ACTIVE_IPMI_TOOL = self.IPMITOOL
+        host_conf_cmd = ""
 
-    def _run_ipmitool_subcommand(self, subcommand, grep_args=None, out_file=subprocess.PIPE):
-        """executes ipmitool sub-commands, and optionally greps the output"""
+        # Set ipmitool to ipmisimtool if activated.
+        if os.path.exists("/tmp/activate_ipmisimtool"):
+            cmd = self.IPMISIMTOOL + " sel info"
+            _, _, retcode = SimpleProcess(cmd).run()
+            if retcode in [0, 2]:
+                self.ACTIVE_IPMI_TOOL = self.IPMISIMTOOL
+                logger.info("IPMI simulator is activated.")
 
-        ipmi_tool = self.IPMITOOL
+        # Fetch channel info from config file and cache.
+        _channel_interface = Conf.get(SSPL_CONF, "%s>%s" %
+                                (BMC_INTERFACE, BMC_CHANNEL_IF))
 
-        # A dummy file path check to select ipmi simulator if
-        # simulator is required, otherwise default ipmitool.
-        if os.path.exists("/tmp/activate_ipmisimtool"):
-            res, retcode = self._run_command(command=f"{self.IPMISIMTOOL} sel info")
-            if retcode == 0:
-                ipmi_tool = self.IPMISIMTOOL
-                logger.info("IPMI simulator is activated")
-
-        command = ipmi_tool + subcommand
-        if grep_args is not None:
-            command += " | grep "
-            if isinstance(grep_args, list):
-                grep_args_str = ""
-                for arg in grep_args:
-                    grep_args_str = "'{}' ".format(arg)
-                command += grep_args_str
-            else:
-                command += "'{}'".format(grep_args)
-        res, retcode = self._run_command(command, out_file)
+        _active_interface = store.get(self.ACTIVE_INTERFACE, None)
+
+        # Set host_conf_cmd based on channel info.
+        if _channel_interface == self.LAN_IF and \
+           _active_interface != self.SYSTEM_IF and \
+           self.ACTIVE_IPMI_TOOL != self.IPMISIMTOOL:
+            bmc_ip = Conf.get(GLOBAL_CONF, BMC_IP_KEY, '')
+            bmc_user = Conf.get(GLOBAL_CONF, BMC_USER_KEY, 'ADMIN')
+            bmc_secret = Conf.get(GLOBAL_CONF, BMC_SECRET_KEY, 'ADMIN')
+
+            decryption_key = encryptor.gen_key(MACHINE_ID,
+                                               ServiceTypes.SERVER_NODE.value)
+            bmc_pass = encryptor.decrypt(decryption_key,
+                                         bmc_secret, self.NAME)
+
+            host_conf_cmd = f"-I lanplus -H {bmc_ip} -U {bmc_user} -P {bmc_pass}"
+
+        # generate the final cmd and execute on shell.
+        command = " ".join([self.ACTIVE_IPMI_TOOL, host_conf_cmd, subcommand])
+
+        out, error, retcode = SimpleProcess([command]).run(shell=True)
```

Comment:
The command is not working without `shell=True`  and  `[]`

Reply:
`shell=True` is discouraged as it is platform dependent and also prone to shell injection. What's the error you are getting with `shell=False`?

Reply:
```
>>> from cortx.utils.process import SimpleProcess
>>> print(SimpleProcess(['/usr/bin/ipmisimtool sdr type "Fan"']).run())
('', 'SubProcess Error: [Errno 2] No such file or directory: \'/usr/bin/ipmisimtool sdr type "Fan"\': \'/usr/bin/ipmisimtool sdr type "Fan"\'', -1)
>>> print(SimpleProcess(['/usr/bin/ipmisimtool sdr type "Fan"']).run(shell=True))
(b'Fan Redundancy   | 0Ch | ok  | 30.1 | Fully Redundant', b'', 0)
>>>
```

Reply:
If you don't want to pass `shell=True`, don't create command string and pass in []. 
Just pass sequence of program arguments.

Ex. 
```
>>> print(SimpleProcess(['echo "ABC"']).run())
('', 'SubProcess Error: [Errno 2] No such file or directory: \'echo "ABC"\': \'echo "ABC"\'', -1)
>>> print(SimpleProcess(['echo "ABC"']).run(shell=True))
(b'ABC\n', b'', 0)
>>> print(SimpleProcess(['echo', '"ABC"']).run())
(b'"ABC"\n', b'', 0)

```

Why it works this way 
From python docs
```args is required for all calls and should be a string, or a sequence of program arguments. Providing a sequence of arguments is generally preferred, as it allows the module to take care of any required escaping and quoting of arguments (e.g. to permit spaces in file names). If passing a single string, either shell must be True (see below) or else the string must simply name the program to be executed without specifying any arguments.```

Reply:
Hi @sandeepsgt , 
I have tried giving cmd in form of arguments but it doesn't work even with shell=True
```
>>> print(SimpleProcess(['/usr/bin/ipmisimtool', 'sdr', 'type',  '"Fan"']).run())
(b'No records found for given sdr or Invalid command - sdr type "Fan". IPMISIMTOOL can\'t process it.\n', b'', 1)

>>> print(SimpleProcess(['/usr/bin/ipmisimtool', 'sdr', 'type',  '"Fan"']).run(shell=True))
(b'', b'', 0)

```
Actually even SimpleProcess is implemented in the same way, It converts a str into a list of args and then pass it for the same reason you explained, but that is not working.
```
>>> sp = SimpleProcess('/usr/bin/ipmisimtool sdr type "Fan"')
>>> o, e, rc = sp.run()
>>> print(sp._cp)
CompletedProcess(args=['/usr/bin/ipmisimtool', 'sdr', 'type', '"Fan"'], returncode=1, stdout=b'No records found for given sdr or Invalid command - sdr type "Fan". IPMISIMTOOL can\'t process it.\n', stderr=b'')
```

I tried multiple combinations and only cmd that worked for me was providing the whole cmd as single arg in list (so that Simple process don't break it into args list), and keeping shell=True
```
>>> print(SimpleProcess(['/usr/bin/ipmisimtool sdr type "Fan"']).run(shell=True))
(b'Fan Redundancy   | 0Ch | ok  | 30.1 | Fully Redundant', b'', 0)
```

Even in the previous implementation we were using shell=True
```
def _run_command(self, command, out_file=subprocess.PIPE):
          """executes commands"""
          process = subprocess.Popen(command, shell=True, stdout=out_file, stderr=subprocess.PIPE)

```

Reply:
Hi, @cdeshmukh, found the solution,
So the cmd needs to properly spllit in `cortx-utils:SimpleProcess` as shown below,
we need to use `split()` method from `shlex` instead of `string` class.
```
>>> import shlex
>>> cmd='/usr/bin/ipmisimtool sdr type "Fan"'
>>> args = shlex.split(cmd)
>>> args
['/usr/bin/ipmisimtool', 'sdr', 'type', 'Fan']
>>> print(SimpleProcess(args).run())
(b'Fan Redundancy   | 0Ch | ok  | 30.1 | Fully Redundant', b'', 0)
``` 
so should I `raise a bug and PR for utils` **OR** `use shlex in sspl and pass the apropriate list`??

Reply:
Thanks for explaining this with patience in call. 
As discussed, lets use shlex split in sspl first & create a bug for SimpleProcess Command implementation also